### PR TITLE
fix: add resource_keys variable to handle dynamic resources

### LINKED ISSUE #76: For Each for the ressource makes problems on not existent project_numbers
### TL;DR

If you deploy a Project in the same code where using this module and the project number is not created yet i getting the error i should use --target.

### Expected behavior

tf plan accept that the project is not exists and create depends on length only the needed count of ressource

### Observed behavior

```
Error: Invalid for_each argument

  on ../terraform-google-vpc-service-controls/modules/regular_service_perimeter/main.tf line 188, in resource "google_access_context_manager_service_perimeter_resource" "service_perimeter_resource":
 188:   for_each       = toset(formatlist("projects/%s", var.resources))

The "for_each" value depends on resource attributes that cannot be determined
until apply, so Terraform cannot predict how many instances will be created.
To work around this, use the -target argument to first apply only the
resources that the for_each depends on.

```

### Terraform Configuration

```hcl
module "org_policy" {
  source      = "terraform-google-modules/vpc-service-controls/google"
  parent_id   =
  policy_name = "test"

}

module "test-vpc-perimeter" {
  source         = "../terraform-google-vpc-service-controls/modules/regular_service_perimeter"
  policy         = module.org_policy.policy_id
  perimeter_name = "test_perimeter"
  description    = "test 123"
  resources      = [module.project2.project_number,module.project.project_number]

}


module "project" {
  source            = "terraform-google-modules/project-factory/google"
  name              = "Test Project AWEB"
  project_id        = "test-project-aweb"
  random_project_id = true
  org_id            = null
  billing_account   = ""
  folder_id         = ""
}

module "project2" {
  source            = "terraform-google-modules/project-factory/google"
  name              = "Test Project AWEB"
  project_id        = "test-project-aweb"
  random_project_id = true
  org_id            = null
  billing_account   = ""
  folder_id         = ""
}
```

### Terraform Version

```sh
Terraform v0.13.7
+ provider registry.terraform.io/hashicorp/google v4.6.0
+ provider registry.terraform.io/hashicorp/google-beta v4.6.0
+ provider registry.terraform.io/hashicorp/null v3.1.0
+ provider registry.terraform.io/hashicorp/random v3.1.0
+ provider registry.terraform.io/hashicorp/time v0.7.2

Your version of Terraform is out of date! The latest version
is 1.1.3. You can update by downloading from https://www.terraform.io/downloads.html
```

### Additional information

I tried to solve it via count but the problem will be then if somebody change the order or remove entry in the middle of the list that triggers recreation of ressources normaly we not wanna touch.

### PR DESCRIPTION
**Pull request by morgante**
Now that resources are created using `for_each`, if the resources are created in the same `apply` they need to have a (separate) key.

Fixes #76.

### CODE REVIEW

**Comment by comment-bot-dev** [comment]
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_documentation
 The documentation needs to be regenerated. Please run `make generate_docs`. 
 ```bash 
Checking for documentation generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/modules/regular_service_perimeter/README.md /tmp/tmp.uBsXtjtXMH/generate_docs/workspace/modules/regular_service_perimeter/README.md
106a107
> | resource\_keys | A list of keys to use for the Terraform state. The order should correspond to var.resources and the keys must not be dynamically computed. If `null`, var.resources will be used as keys. | `list(string)` | `null` | no |
Error: Documentation generation has not been run, please run the
'make docker_generate_docs' command and commit the above changes.
 ``` 
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
modules/regular_service_perimeter/main.tf
--- old/modules/regular_service_perimeter/main.tf
+++ new/modules/regular_service_perimeter/main.tf
@@ -186,7 +186,7 @@
 locals {
   resource_keys = var.resource_keys != null ? var.resource_keys : var.resources
   resources = {
-    for rk in local.resource_keys:
+    for rk in local.resource_keys :
     rk => var.resources[index(local.resource_keys, rk)]
   }
 }
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
 ```

File: examples/simple_example_dynamic/outputs.tf [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+output "policy_name" {
+  value = var.policy_name
```
Comment (bharathkkb) [inline examples/simple_example_dynamic/outputs.tf:18]: nit: policy id maybe more useful?

File: examples/simple_example_dynamic/versions.tf [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+terraform {
+  required_version = ">= 0.12"
```
Comment (bharathkkb) [inline examples/simple_example_dynamic/versions.tf:18]:
```suggestion
  required_version = ">= 0.13"
```

File: examples/simple_example_dynamic/main.tf [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  source         = "../../modules/bridge_service_perimeter"
+  policy         = module.access_context_manager_policy.policy_id
+  perimeter_name = "bridge_perimeter_1"
+  description    = "Some description"
+
+  resources = concat(
+    module.regular_service_perimeter_1.shared_resources["all"],
+    module.regular_service_perimeter_2.shared_resources["all"],
+  )
+  resource_keys = ["one", "two", "three"]
```
Comment (bharathkkb) [inline examples/simple_example_dynamic/main.tf:33]: This is slightly confusing as its not apparent `module.regular_service_perimeter_2.shared_resources["all"]` is a list of two elems and hence there are three resource_keys.  Can we add a comment?

**Review by bharathkkb — APPROVED** [review]

File: examples/simple_example_dynamic/main.tf (hunk 3 above)
Comment (morgante) [reply examples/simple_example_dynamic/main.tf]: Switched to using explicit list for resources to make this clearer.

File: examples/simple_example_dynamic/outputs.tf (hunk 1 above)
Comment (morgante) [reply examples/simple_example_dynamic/outputs.tf]: Switched.

**Pull request merged**